### ISSUE TITLE
Is it possible to add time rule to part of content in a chunk?

### ISSUE THREAD

**Issue by dzhuang** [inducer/relate#109]
I want to split my course materials into chunks by chapters, not by lectures. For A long chapter which can last 4 lectures, I want to add time rule, like ``if_after lecture 3`` then ``show: True``,  in ``course.yml`` so that some text in that chuck are hidden until the beginning of lecture 3?

**Comment by inducer** [comment]
No, a chunk is the smallest granularity at the moment. Anything smaller you'd probably have to hack together out of Jinja macros (that could be built in).

**Comment by dzhuang** [comment]
Following your suggestion, my idea now is to create a jinja macro named ``show_if_after`` and use javascript to control the visibility of some blocks. However, I have no idea how to pass the ``Event`` list, as well as ``get_now_or_fake_time``, as context to the macro, because I don't want to use ISO format datetime. Any suggestions? Thanks.

**Comment by dzhuang** [comment]
Seem adding a ``jinjia_env`` parameter to ``get_processed_course_chunks`` will do the work, and then validation should modified, is that correct?

**Comment by inducer** [comment]
No. I would introduce a Jinja macro `{% show_if_before "lecture 17 + 3 days" %}` that you can then use to your heart's content *within* each chunk.

One wrinkle with time-based macro expansions is that the resulting markup is no longer cacheable.

**Comment by dzhuang** [comment]
Yes, I am trying to introduce that macro. However, I still have no idea how to pass existing Events and now_datetime to that macro. If I don't do that, how can the macro understand ``lecture 17 + 3 days``?

**Comment by inducer** [comment]
https://github.com/inducer/relate/blob/c204d84163d9973a826b483aacf37ae3230a0e31/course/content.py#L566

`markup_to_html` already knows the course, which is where you'd inject your macro. From there, you can use `parse_date_spec` to parse a date like `lecture 17 + 3 days`.

**Comment by dzhuang** [comment]
Thank you, I followed your suggestions and successfully parsed event datetime, which works both for  ``chunks`` and for ``pages``. However, to get the ``now_datetime`` (using ``get_now_or_fake_time``), we need to have a ``request`` object, and that seems can't be done by only hacking the ``markup_to_html`` method.

**Comment by inducer** [comment]
Right--we would need to make ways all throughout RELATE to pass around `now_datetime` into `markup_to_html`.

**Comment by inducer** [comment]
It would perhaps be smart to package up all this information into a `MarkupContext` (or some such) object, so that other pieces of info can be added more easily in the future.

**Comment by dzhuang** [comment]
Tried to fix by https://github.com/inducer/relate/pull/115.
